### DISCUSSION TITLE
Running Directus 9 over HTTPS

### DISCUSSION THREAD

**Discussion by EPMatt** [directus/directus#3307] [Archive]
Hi there,
I've just installed a brand new Directus 9 installation.
How can I configure the server to run over HTTPS?

**Comment by rijkvanzanten** [comment]
Heya!

This totally depends on where / how you intend to host Directus. When you use a service like DigitalOcean App Platform, Heroku, or Platform.sh, domains and SSL are managed ootb. When hosting it "manually" on a EC2 instance/Droplet/VPS, you can rely on something like Nginx or Caddy for a reverse proxy.

This is a good resource on how to get that going: https://www.digitalocean.com/community/tutorials/how-to-set-up-a-node-js-application-for-production-on-ubuntu-16-04

**Reply by EPMatt** [reply]
Thank you very much!
I'd also suggest adding this answer as a little note in the docs just to help newbies in completing the ssl configuration. 👍🏻

**Reply by efpage** [reply]
Installed Directus9 via cli. Defaults to http://myPlatform:8055. How enable https? Tried via Apache, but did not manage to get things runnings. Any hints welcome.

**Reply by joselcvarela** [reply]
Hello @efpage 

On my side, to make this happen I use docker with [traefik](https://doc.traefik.io/traefik/v1.7/) combination.
Here is an example: https://doc.traefik.io/traefik/user-guides/docker-compose/basic-example/

But this goes beyond Directus scope, so you can search on google or youtube for: "node nginx letsencrypt"
Node is what you need to be installed to run Directus.
Nginx is used for reverse proxy, i.e. redirect traffic. In this case you would need it to redirect from 443 port to 8055 and configure proper certificates.
Let's encrypt is a service which can provide you valid certificates and renewal with ease.

**Reply by efpage** [reply]
Hy joselcvarela, thank you for your explanation. Sure this is beyond the Directus-scope, but as anybody has to run the service as https it is worth to mention. Apache or nginx configuration is not that easy. 

I have a server running with Apache2 and Letsencrypt certificates.  Here /etc/apache2/sites-enabled/000-default-le-ssl.conf we find the ssl-configuration:

```
<IfModule mod_ssl.c>
<VirtualHost *:443>

        ... 
           some ssl stuff 
        ...

	# directus
	ProxyRequests Off
	ProxyPreserveHost On
	RewriteRule "/directus$" "https://myserver.com/directus/admin/index.html" [R]
	ProxyPass /directus/ http://myserver.com:8055/
        ProxyPassReverse /directus/ http://myserver.com:8055/

</VirtualHost>
</IfModule>
```
This seems to work, so I can access directus via myserver.com/directus, but the RewriteRule makes some trouble still.

PS: Found that WYSIWYG-Editor is not showing up with this configuration.

**Reply by joselcvarela** [reply]
Have you solved @efpage? I can provide an example using docker-compose and traefik, if you want.

**Reply by rijkvanzanten** [reply]
@efpage An even easier solution might be to use Caddy (https://caddyserver.com) with the config file:

```
mydomain.com {
    reverse_proxy localhost:8055
}
```

**Reply by efpage** [reply]
It seems, TinyMCE (which is used as a WYSIWYG-Editor) needs some additional proxy settings to work... Any suggestions welcome

**Reply by lerogit** [reply]
> Have you solved @efpage? I can provide an example using docker-compose and traefik, if you want.

Hello,

Would you mind sharing an example using docker-compose and traefik? I'm having a hard trying to make this setup work.

**Reply by jessydm** [reply]
> @efpage An even easier solution might be to use Caddy (https://caddyserver.com) with the config file:
> 
> ```
> mydomain.com {
>     reverse_proxy localhost:8055
> }
> ```

Easiest setup.

**Reply by teichsta** [reply]
> Have you solved @efpage? I can provide an example using docker-compose and traefik, if you want.

@joselcvarela i would be very interested in this :), thanks guys!

**Comment by biapar** [comment]
Then a reverse proxy is obligatory. Right?

**Reply by joselcvarela** [reply]
Hey @biapar 
Yes, it's an hard work we may would not be willing to do.
Although you can open an `Idea` here in Discussions for support it, so others can upvote.

**Reply by efpage** [reply]
You cannot access directus from an https-page as long as it is using http. So, anybody who wants to use directus has to solve this. So, should we state that directus currently is not usable?

**Reply by biapar** [reply]
> You cannot access directus from an https-page as long as it is using http. So, anybody who wants to use directus has to solve this. So, should we state that directus currently is not usable?

So Cannot we use reverse proxy?

**Reply by efpage** [reply]
I´m not a server admin (like most directus users probably are), bis as far as I can see from the apache-documentation , I HAVE enabled the reverse proxy:

	ProxyPass /directus/ http://myserver.com:8055/
        ProxyPassReverse /directus/ http://myserver.com:8055/

Nevertheless tinyMCE seems to need some more?!? There are 10.000 possibilities to do things wrong on apache, but only one is right, so there is definitvely some need for advice in this point. 

But maybe anybody is just playing around with directus and not using it in a real life case, so maybe I´m the only one that needs to solve this problem...

**Reply by rijkvanzanten** [reply]
> You cannot access directus from an https-page as long as it is using http. So, anybody who wants to use directus has to solve this. So, should we state that directus currently is not usable?

No, that's false. You'd use a reverse proxy like Nginx or a load balancer to route the traffic and enable HTTPS 👍🏻

**Reply by rijkvanzanten** [reply]
> > You cannot access directus from an https-page as long as it is using http. So, anybody who wants to use directus has to solve this. So, should we state that directus currently is not usable?
> 
> 
> 
> So Cannot we use reverse proxy? 

You can use any reverse proxy you'd like 👍🏻

**Reply by biapar** [reply]
@rijkvanzanten a simple NGNIX config
server {
    # Listen HTTP
    listen 80;
    server_name api.example.com;

    # Redirect HTTP to HTTPS
    return 301 https://$host$request_uri;
}

server {
    # Listen HTTPS
    listen 443 ssl;
    server_name api.example.com;

    # SSL config
    ssl_certificate /path/to/your/certificate/file;
    ssl_certificate_key /path/to/your/certificate/key;

    # Proxy Config
    location / {
        proxy_pass http://directus;
        proxy_http_version 1.1;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Server $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $http_host;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass_request_headers on;
    }
}

**Reply by efpage** [reply]
Yes, quite simple, just what I expected. But I have Apache running, possibly even simpler? Should be a blink of an eye for someone who was born on plant linux. So, it should be quite simple to tell me what´s wrong:

As my Apache is cofigured to redirect any http-request to https, I cannot use the base path. So I set up the reverse proxy like this (among other tools running on this server):

	# directus
	ProxyPass /directus http://12.34.56.78:8055
	ProxyPassReverse /directus http://12.34.56.78:8055

Just, what i get ist:

If I try 

https://myServer/directus

It switches to 

https://myServer/admin

not nice. If I use 

https://myServer/directus/admin

nothing happens.

So, what is the simple reason?

**Reply by rijkvanzanten** [reply]
@efpage most likely that your PUBLIC_URL environment variable isn't set to `https://myServer/directus`

**Reply by efpage** [reply]
That´s it! Now it ~~walks~~ works! Took me some hours to find out, that you cannot run directus in a subpath without some tweaks. Thank you very much for your support!
![This is an image](https://synapsenstau.de/wp-content/uploads/2021/03/Bewegung-unterbewusste-interpretieren.gif)

**Reply by efpage** [reply]
Sorry to reach out agin, but one issue still there. I use this url:

PUBLIC_URL="https://myserver.xx/directus"

If I call the adress "https://myserver.xx/directus it redicets to 
"https://myserver.xx/admin

I need to manually use "https://myserver.xx/directus/admin to get into the admin panel. Any suggestion?

**Reply by azrikahar** [reply]
@efpage likely configuring `ROOT_REDIRECT` to `./directus/admin` (currently it's `./admin` by default) should do the trick for you. Reference: https://docs.directus.io/configuration/config-options/#general

**Reply by efpage** [reply]
Yes, it works. But documentation could definitively be more clear it in this point! Even the hints for PUBLIC_URL are a bit misleading:

"The PUBLIC_URL value is used for things like OAuth redirects, forgot-password emails, and logos that needs to be publicly available on the internet."

**Reply by azrikahar** [reply]
Hmm technically the `PUBLIC_URL` is correct here (at least from my understanding), it's just that `ROOT_REDIRECT` being the one "disrupting" this description which is understandable. If you have any suggestions or perhaps even a PR to refine/clarify this, it'd be much appreciated! 😊

**Reply by efpage** [reply]
Setting up a new system is always an adventure. There are thousands of options to make it wrong, but only one way is correct. I stepped inito a pitfall a I wanted to set up directus on my test server where other things are already running, so each application gets it´s own path like www.myserver.com/directus or www.myserver.com/strapi. 

As you need to install a 433-redirect anyway, this is not an uncommon situation. It´s just not obvious enough that you need some tweaks to get this running.

Technically I cannot help much in this case.

But thanks for the help anyway,
BR Eckehard
